src/Servers/Kestrel/test/InMemory.FunctionalTests/MaxRequestBodySizeTests.cs

```diff
@@ -60,6 +60,89 @@ public async Task RejectsRequestWithContentLengthHeaderExceedingGlobalLimit()
         Assert.Equal(CoreStrings.FormatBadRequest_RequestBodyTooLarge(globalMaxRequestBodySize), requestRejectedEx.Message);
     }
 
+    [Fact]
+    public async Task RejectsRequestWithBodySizeExceedingPerRequestLimitAndExceptionWasCaughtByApplication()
+    {
+        var maxRequestBodySize = 10;
+        await using (var server = new TestServer(async context =>
+        {
+#pragma warning disable CS0618 // Type or member is obsolete
+            BadHttpRequestException requestRejectedEx = await Assert.ThrowsAsync<BadHttpRequestException>(async () =>
+#pragma warning restore CS0618 // Type or member is obsolete
+            {
+                using (var stream = new StreamReader(context.Request.Body))
+                {
+                    string body = await stream.ReadToEndAsync();
+                }
+            });
+            context.Response.StatusCode = 413;
+            await context.Response.WriteAsync("");
+            throw requestRejectedEx;
+        },
+        new TestServiceContext(LoggerFactory) { ServerOptions = new KestrelServerOptions { Limits = { MaxRequestBodySize = maxRequestBodySize } } }))
+        {
+            using var connection = server.CreateConnection();
+            await connection.Send(
+                "POST / HTTP/1.1",
+                "Host:",
+                "Content-Length: " + (maxRequestBodySize + 1),
+                "",
+                new string('a', maxRequestBodySize + 1));
+            await connection.ReceiveEnd(
+                "HTTP/1.1 413 Payload Too Large",
+                "Connection: close",
+                $"Date: {server.Context.DateHeaderValue}",
+                "Server: Kestrel",
```

Comment:
I got the `Server: kestrel` header in the test but didn't find it in the others tests, is normal to get this header?

Reply:
I think this is because you're doing `ServerOptions = new KestrelServerOptions { Limits = ...` instead of `ServerOptions = { Limits = ...`. The former overwrites `ServerOptions.AddServerHeader` to its default value of `true` instead of keeping it `false` like we do for most tests [here](https://github.com/dotnet/aspnetcore/blob/a364bcbd1c360762d5e8bda8fe04fa004973489a/src/Servers/Kestrel/shared/test/TestServiceContext.cs#L43).